cli/command/telemetry.go

```diff
@@ -0,0 +1,214 @@
+	exp, err := dockerMetricExporter(ctx, cli)
+	if err != nil {
+		otel.Handle(err)
+	}
+	defaultOpts := []sdkmetric.Option{
+		sdkmetric.WithResource(cli.Resource()),
+		sdkmetric.WithReader(newCLIReader(exp)),
+	}
+	return sdkmetric.NewMeterProvider(append(defaultOpts, opts...)...)
+}
+
+// WithResourceOptions configures additional options for the default resource. The default
+// resource will continue to include its default options.
+func WithResourceOptions(opts ...resource.Option) CLIOption {
+	return func(cli *DockerCli) error {
+		cli.res.AppendOptions(opts...)
+		return nil
+	}
+}
+
+// WithResource overwrites the default resource and prevents its creation.
+func WithResource(res *resource.Resource) CLIOption {
+	return func(cli *DockerCli) error {
+		cli.res.Set(res)
+		return nil
+	}
+}
+
+type telemetryResource struct {
+	res  *resource.Resource
+	opts []resource.Option
+	once sync.Once
+}
+
+func (r *telemetryResource) Set(res *resource.Resource) {
+	r.res = res
+}
+
+func (r *telemetryResource) Get() *resource.Resource {
+	r.once.Do(r.init)
+	return r.res
+}
+
+func (r *telemetryResource) init() {
+	if r.res != nil {
+		r.opts = nil
+		return
+	}
+
+	opts := r.defaultOptions()
+	if len(r.opts) > 0 {
+		opts = append(opts, r.opts...)
+	}
+
+	res, err := resource.New(context.Background(), opts...)
+	if err != nil {
+		otel.Handle(err)
+	}
+	r.res = res
+
+	// Clear the resource options since they'll never be used again and to allow
+	// the garbage collector to retrieve that memory.
+	r.opts = nil
+}
+
+func (r *telemetryResource) defaultOptions() []resource.Option {
+	return []resource.Option{
+		resource.WithDetectors(serviceNameDetector{}),
+		resource.WithAttributes(
+			// Use a unique instance id so OTEL knows that each invocation
+			// of the CLI is its own instance. Without this, downstream
+			// OTEL processors may think the same process is restarting
+			// continuously.
+			semconv.ServiceInstanceID(uuid.Generate().String()),
+		),
+		resource.WithFromEnv(),
+		resource.WithTelemetrySDK(),
+	}
+}
+
+func (r *telemetryResource) AppendOptions(opts ...resource.Option) {
+	if r.res != nil {
+		return
+	}
+	r.opts = append(r.opts, opts...)
+}
+
+type serviceNameDetector struct{}
+
+func (serviceNameDetector) Detect(ctx context.Context) (*resource.Resource, error) {
+	return resource.StringDetector(
+		semconv.SchemaURL,
+		semconv.ServiceNameKey,
+		func() (string, error) {
+			return filepath.Base(os.Args[0]), nil
+		},
+	).Detect(ctx)
+}
+
+// cliReader is an implementation of Reader that will automatically
+// report to a designated Exporter when Shutdown is called.
+type cliReader struct {
+	sdkmetric.Reader
+	exporter sdkmetric.Exporter
+}
+
+func newCLIReader(exp sdkmetric.Exporter) sdkmetric.Reader {
+	reader := sdkmetric.NewManualReader(
+		sdkmetric.WithTemporalitySelector(deltaTemporality),
+	)
+	return &cliReader{
+		Reader:   reader,
+		exporter: exp,
+	}
+}
+
+const shutdownTimeout = 2 * time.Second
```

Comment:
2 seconds is a while – especially if we're talking about really quick invocations – think `docker exec my-container true` which normally exits somewhere in the range of a few milliseconds. If this would block the CLI from exiting, I'd reduce this significantly.

Reply:
So my knowledge about this is super limited, but here (in the CLI process) we'd be sending stuff to the collector (which generally is going to be running on a local machine), right? Not sending stuff over the internet? So I would think it would be way faster than 200ms generally, but I could be wrong.

Reply:
The collector should be running in DD, so I'd also _expect_ it to be faster than 200ms. 
But TBH I haven't looked at any of the collector impls so far

Reply:
It should be much faster than 200ms as it would be running locally. I'd expect a couple of milliseconds at most, but I mostly want to set the timeout to whatever would be the most permissive that doesn't interfere with a user using the tool just to ensure we don't accidentally set it too low.

Reply:
Right, that's what I thought. In that case, would you feel okay about setting this at 50ms? Thinking about the edge cases where some weird condition happens with the collector and it doesn't immediately fails but hangs, in that case 50ms is unnoticeable but two seconds is VERY noticeable.

Reply:
Agreed, this needs to be much shorter; high-latency connections should be something of an error state, and this will make them more painful.

Reply:
Done. I also realized the timeout didn't apply to traces and found the option to apply this timeout to traces too. It's been renamed to `exportTimeout` when I end up pushing the change.